packages/runtime-vapor/__tests__/componentProps.spec.ts

```diff
@@ -0,0 +1,399 @@
+    // expect(defaultFn).toHaveBeenCalledTimes(1) // FIXME: failed
+
+    render(
+      Comp as any,
+      {
+        get bar() {
+          return { b: 4 }
+        },
+      },
+      host,
+    )
+    expect(props.foo).toBe(1)
+    expect(props.bar).toEqual({ b: 4 })
+    // expect(defaultFn).toHaveBeenCalledTimes(1) // FIXME: failed
+  })
+
+  test('using inject in default value factory', () => {
+    // FIXME: is it necessary?
+  })
+
+  test('props type support BigInt', () => {
+    const Comp = {
+      props: {
+        foo: BigInt,
+      },
+      render() {
+        const instance = getCurrentInstance()!
+        const t0 = template('<div></div>')
+        const n0 = t0()
+        const {
+          0: [n1],
+        } = children(n0)
+        watchEffect(() => {
+          setText(n1, instance.props.foo)
+        })
+        return n0
+      },
+    }
+
+    render(
+      Comp as any,
+      {
+        get foo() {
+          return (
+            BigInt(BigInt(100000111)) + BigInt(2000000000) * BigInt(30000000)
+          )
+        },
+      },
+      '#host',
+    )
+    expect(host.innerHTML).toBe('<div>60000000100000111</div>')
+  })
+
+  // #3288
+  // test('declared prop key should be present even if not passed', async () => {
+  // let initialKeys: string[] = []
+  // const changeSpy = vi.fn()
+  // const passFoo = ref(false)
+  // const Comp = {
+  //   props: ['foo'],
+  //   setup() {
+  //     const instance = getCurrentInstance()!
+  //     initialKeys = Object.keys(instance.props)
+  //     watchEffect(changeSpy)
+  //     return {}
+  //   },
+  //   render() {
+  //     return {}
+  //   },
+  // }
+  // const Parent = createIf(
+  //   () => passFoo.value,
+  //   () => {
+  //     return render(Comp as any, { foo: 1 }, host) // TODO: createComponent fn
+  //   },
+  // )
+  // // expect(changeSpy).toHaveBeenCalledTimes(1)
+  // })
+
+  // #3371
+  // test(`avoid double-setting props when casting`, async () => {
+  //   // TODO: proide, slots
+  // })
+
+  test('support null in required + multiple-type declarations', () => {
+    const Comp = {
+      props: {
+        foo: { type: [Function, null], required: true },
+      },
+      render() {},
+    }
+
+    expect(() => {
+      render(Comp as any, { foo: () => {} }, host)
+    }).not.toThrow()
+
+    expect(() => {
+      render(Comp as any, { foo: null }, host)
+    }).not.toThrow()
+  })
+
+  // // #5016
+  // test('handling attr with undefined value', () => {
+  //   // TODO: attrs
+  // })
+
+  // #691ef
+  test('should not mutate original props long-form definition object', () => {
+    const props = {
+      msg: {
+        type: String,
+      },
+    }
+    const Comp = defineComponent({
+      props,
+      render() {},
+    })
+
+    render(Comp as any, { msg: 'test' }, host)
+
+    expect(Object.keys(props.msg).length).toBe(1)
+  })
+
+  // NOTE: not supported
+  // optimized props update
+  // mixins
+  // validator
+  // warn
+  // caching
```

Comment:
Is my understanding correct about this area?

Reply:
for `optimized props update`: maybe it's unnecessary (I'm not sure for now). But Vapor can also pass the test. So let's add it first and then remove it if it's useless at the end of the development of Vapor.

Reply:
`validator` should be implemented as well. Let's add it as `test.todo`

Reply:
`warn`: What the reason is that you think it should be not supported? IMHO it should be also included.

Reply:
For the rest, let's skip them.